### ISSUE TITLE
QueryContext blocks when the network is down

### ISSUE THREAD

**Issue by tiwo** [microsoft/go-mssqldb#160]
# QueryContext blocks when the network is down

I want to deal with sometimes slow or unreliable network connections, and use QueryContext to limit query time.

After a connection is established (by Ping() or previous queries), and then fails (in my experiments I disconnect the VPN connection to the server), the next QueryContext hangs indefinitely even after the context is cancelled.

To reproduce, see https://github.com/tiwo/2023-gomssqldb-issue-with-timeout

    % CONNECTIONSTRING='odbc:DRIVER={ODBC Driver 17 for SQL Server};SERVER={10.0.11.13};PORT={11433};DATABASE={ABC};USER ID={johndoe};PASSWORD={secret password};ApplicationIntent=ReadOnly;' \
    go run main.go
    03:10:39 main.go:68: main()
    03:10:39 main.go:31: query_with_timeout("SELECT 'first';")
    03:10:41 main.go:62: query_with_timeout returning "first"
    03:10:41 main.go:91: to reproduce, disconnect from your network now
    03:10:46 main.go:31: query_with_timeout("SELECT 'second';")
    03:10:58 main.go:38: Cancelling query... state of ctx is: context.deadlineExceededError{})
    03:11:29 main.go:84: Closing database
    ^C03:17:57 main.go:75: Received SIGINT; exiting.

Tested with Go 1.21.3, go-mssqldb 1.6.0.
For what it's worth, I am setting db.SetConnMaxLifetime which should be unrelated, and I am making sure the context really is cancelled.

Is the context even used in [Stmt.sendQuery](https://github.com/microsoft/go-mssqldb/blob/e51fa150588f719f052ab5b715cc595f94088bc3/mssql.go#L506), except for logging?

**Comment by shueybubbles** [comment]
Is this the same problem as #91?
PR #109 attempts to fix that but is failing all the tests.

**Comment by tiwo** [comment]
no, not quite the same. I think #91 actually describes expected behaviour, not a bug. After all, Ping() and Query() _must_ block as long as the "pong" is yet to be seen and the TCP connection is still alive (which is usually well after the network cable is disconnected). The solution to issue #91 is "replace Ping by PingContext to avoid the halting problem", I think.

But for PingContext and QueryContext it is different, I would expect for them to abort once the context is done, "no matter what". They shouldn't wait for a slow network. It _is_ the same loop that blocks though. So this present issue (and the identical denisenkom/go-mssqldb#527) needs some unblocking.

I'm a bit out of my depth. My best lead is this: [readCancelConfirmation](https://github.com/microsoft/go-mssqldb/blob/670fd5877e155774d91c0861f438e7eef72e59f8/token.go#L1279C1-L1279C61) would need to (receive as a parameter and) consider the context, and stop blocking there. Maybe the connection could just be discarded and closed. Or the incoming tokens up until the cancellation confirmation would have to be consumed _in the background_.

**Comment by navigator-dev** [comment]
Is there any update on this issue? Is there a current workaround, like moving all queries into goroutines? That's a bit tedious and I'm unsure if doing so would mean we get memory leaks.

**Comment by vasily-kirichenko** [comment]
@tiwo 

I'm afraid we should cast to `doneStruct`, not to the pointer to it:

```go
confirmToken, isConfirmation := tok.(doneStruct)
```

This is because the original function is:

```go
func readCancelConfirmation(tokChan chan tokenStruct) bool {
	for tok := range tokChan {
		switch tok := tok.(type) {
		default:
		// just skip token
		case doneStruct: // the struct type, not the pointer to it
			if tok.Status&doneAttn != 0 {
				// got cancellation confirmation, exit
				return true
			}
		}
	}
	return false
}
```
